velox/expression/tests/ExprCompilerTest.cpp

```diff
@@ -0,0 +1,144 @@
+    return std::make_shared<core::CallTypedExpr>(BOOLEAN(), args, "or");
+  }
+
+  core::TypedExprPtr call(
+      const std::string& name,
+      const std::vector<core::TypedExprPtr>& args) {
+    std::vector<TypePtr> argTypes;
+    for (auto& arg : args) {
+      argTypes.push_back(arg->type());
+    }
+    auto returnType = parse::resolveScalarFunctionType(name, argTypes);
+    return std::make_shared<core::CallTypedExpr>(returnType, args, name);
+  }
+
+  core::TypedExprPtr bigint(int64_t value) {
+    return std::make_shared<core::ConstantTypedExpr>(value);
+  }
+
+  std::function<core::FieldAccessTypedExprPtr(const std::string& name)>
+  makeField(const RowTypePtr& rowType) {
+    return [&](const std::string& name) -> core::FieldAccessTypedExprPtr {
+      auto type = rowType->findChild(name);
+      return std::make_shared<core::FieldAccessTypedExpr>(type, name);
+    };
+  }
+
+  std::unique_ptr<ExprSet> compile(const core::TypedExprPtr& expr) {
+    return std::make_unique<ExprSet>(
+        std::vector<core::TypedExprPtr>{expr}, execCtx_.get());
+  }
+
+  std::shared_ptr<core::QueryCtx> queryCtx_{core::QueryCtx::createForTest()};
+  std::unique_ptr<core::ExecCtx> execCtx_{
+      std::make_unique<core::ExecCtx>(pool_.get(), queryCtx_.get())};
+};
+
+TEST_F(ExprCompilerTest, constantFolding) {
+  auto rowType = ROW({"a"}, {BIGINT()});
+
+  auto field = makeField(rowType);
+
+  // a + (1 + 5) => a + 6
+  auto expression =
+      call("plus", {field("a"), call("plus", {bigint(1), bigint(5)})});
+  ASSERT_EQ("plus(a, 6:BIGINT)", compile(expression)->toString());
+
+  // (a + 1) + 5 is not folded.
+  expression = call("plus", {call("plus", {field("a"), bigint(1)}), bigint(5)});
+  ASSERT_EQ(
+      "plus(plus(a, 1:BIGINT), 5:BIGINT)", compile(expression)->toString());
+}
+
+TEST_F(ExprCompilerTest, andFlattening) {
+  auto rowType =
+      ROW({"a", "b", "c", "d"}, {BOOLEAN(), BOOLEAN(), BOOLEAN(), BOOLEAN()});
+
+  auto field = makeField(rowType);
+
+  auto expression =
+      andCall(field("a"), andCall(field("b"), andCall(field("c"), field("d"))));
+  ASSERT_EQ("and(a, b, c, d)", compile(expression)->toString());
+
+  expression =
+      andCall(andCall(andCall(field("a"), field("b")), field("c")), field("d"));
+  ASSERT_EQ("and(a, b, c, d)", compile(expression)->toString());
+
+  expression =
+      andCall(andCall(field("a"), field("b")), andCall(field("c"), field("d")));
+  ASSERT_EQ("and(a, b, c, d)", compile(expression)->toString());
+
+  expression =
+      orCall(field("a"), andCall(field("b"), andCall(field("c"), field("d"))));
+  ASSERT_EQ("or(a, and(b, c, d))", compile(expression)->toString());
+
+  // Verify no flattening happens when AND is mixed with OR.
+  expression =
+      andCall(field("a"), orCall(field("b"), andCall(field("c"), field("d"))));
+  ASSERT_EQ("and(a, or(b, and(c, d)))", compile(expression)->toString());
+}
+
+TEST_F(ExprCompilerTest, orFlattening) {
```

Comment:
this reminds me of the discussion that it would be nice to support flattening for other types of associative functions as well, like concat().

Reply:
I started working on that: #2551